contracts/helpers/RangeAmountCalculator.sol

```diff
@@ -0,0 +1,62 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.11;
+
+import "../libraries/Math.sol";
+
+/**
+ * Range limit order is used to sell an asset within a given price range.
+ * For example, right now ETH is worth 3000 DAI and you believe that within the next week the price of ETH will rise and reach at least 4000 DAI.
+ * In this case, you can create an ETH -> DAI limit order with a price range of 3000 -> 4000.
+ * Let's say you created a similar order for the amount of 10 ETH.
+ * Someone can file the entire limit at once at an average price of 3500 DAI.
+ * But it is also possible that the limit-order will be filed in parts.
+ * First, someone buys 1 ETH at the price of 3050 DAI, then another 1 ETH at the price of 3150 DAI, and so on.
+ */
+contract RangeAmountCalculator {
+
+    function getRangeTakerAmount(
+        uint256 priceStart,
+        uint256 priceEnd,
+        uint256 totalLiquidity,
+        uint256 fillAmount,
+        uint256 filledFor
+    ) public pure returns(uint256) {
+        unchecked {
+            uint256 remainingLiquidity = totalLiquidity - filledFor;
+            uint256 filledForAfterFill = filledFor + fillAmount;
+            uint256 remainingLiquidityAfterFill = remainingLiquidity - fillAmount;
+
+            uint256 amountBeforeFill = priceStart * remainingLiquidity + priceEnd * filledFor;
+            uint256 amountAfterFill = priceStart * remainingLiquidityAfterFill + priceEnd * filledForAfterFill;
+            uint256 price = (amountBeforeFill + amountAfterFill) / totalLiquidity / 2;
+
+            return fillAmount * price;
+        }
+    }
+
+    function getRangeMakerAmount(
```

Comment:
I can optimize it as:
```
function getRangeMakerAmount(
        uint256 priceStart,
        uint256 priceEnd,
        uint256 totalLiquidity,
        uint256 takingAmount,
        uint256 filledFor
    ) public pure returns(uint256) {
        unchecked {
            uint256 priceRangeDiff = priceEnd - priceStart;
            uint256 filledForSqr = filledFor * filledFor;

            uint256 a = filledFor * (totalLiquidity * (priceRangeDiff) * (takingAmount / filledFor + priceStart) - priceStart * filledFor * priceEnd);
            uint256 b = priceEnd * priceEnd * filledForSqr + priceStart * priceStart * (filledForSqr + totalLiquidity * totalLiquidity);
            uint256 c = priceStart * (totalLiquidity - filledFor) + priceEnd * filledFor;

            return (Math.sqrt(2 * a + b) - c) / priceRangeDiff;
        }
    }
```

The original code used 34853 gas, optimized 34810. (for call: 3000,4000,10,3150,1)

Reply:
300000,400000,15000,32150,150

Original: 45808 gas
Optimized: 45851 gas